### PR TITLE
Eliminate Phases in favor of smarter folds.

### LINKED ISSUE #414: planner: surface failures in logical plan phases related to unbound variables somehow
Currently any "bound" phase can only be applied to plans that include no unbound `Free` instances. If any occur, an exception is thrown (and not handled well). This isn't happening at the moment only because the compiler never produces such a plan.

At a minimum, these errors should be captured, probably using `-\/`, But it's not clear how this would fit into the rest of the Monad stacking that's going on.

Furthermore, it might be useful for these phases to be able to handle "fragments" of plans that have these unbound variables in them, say by supplying a function to handle any unbound variable.

### PR DESCRIPTION
**Pull request by sellout**
* Generalize folds a lot, much more akin to @ekmett’s recursion-schemes
  than Fixplate at this point;
* remove Phase in favor of composing folds directly, eliminating
  intermediate data structures and the problems with rewriting bindings;
* replace Binder with a simpler version for use with boundCata, etc.;
* convert JsExprPhase to “Partial” style (same approach used for SelectorPhase);
* fix jsExpr1 to not require State (and therefore, projectIndex, as well);
* restrain State in workflowƒ (née WorkflowPhase); and
* incidentally, fixes #414.

Shortcomings:

* `boundCata`, etc. are implemented directly, rather than via the
  generalized folds and
* kleisli folds (EG, `lpParaZygoHistoM`) are currently implemented on
  the specific functor rather than using even Binder (hopefully this can
  be fixed with some small changes to Binder).

### CODE REVIEW

**Comment by sellout** [comment]
Well … I was afraid of that. I added a bunch of stuff to fixplate that I ended up not being able to use (yet). And I didn’t write any new tests for it. There may be a few other places, I’ll have to check the report.

But, I think I may just roll back those fixplate changes/additions for now.

File: core/src/main/scala/slamdata/engine/analysis/EnvT.scala [hunk 1]
```diff
@@ -0,0 +1,62 @@
+package slamdata.engine.analysis
+
+import scalaz._
+import Scalaz._
+import Id._
+
+/**
+ * This is the transformer for the (,) comonad.
+ */
+final case class EnvT[E, W[_], A](run: (E, W[A])) { self =>
```
Comment (jdegoes) [inline core/src/main/scala/slamdata/engine/analysis/EnvT.scala]: This should most likely be in the `fp` package.

Comment (sellout) [inline core/src/main/scala/slamdata/engine/analysis/EnvT.scala]: Ah, yeah. I’d like to submit it to Scalaz at some point, but that is a better place in the meantime.

File: core/src/main/scala/slamdata/engine/analysis/fixplate.scala [hunk 2]
```diff
@@ -10,10 +10,10 @@ import Id.Id
 import slamdata.engine.{RenderTree, Terminal, NonTerminal}
 
 sealed trait term {
-  case class Term[F[_]](unFix: F[Term[F]]) {
-    def cofree(implicit f: Functor[F]): Cofree[F, Unit] =
-      Cofree(Unit, Functor[F].map(unFix)(_.cofree))
+  def unzipF[F[_]: Functor, A, B](f: F[(A, B)]): (F[A], F[B]) =
```
Comment (jdegoes) [inline core/src/main/scala/slamdata/engine/analysis/fixplate.scala]: Ditto for this, if it doesn't already exist in Scalaz, which I'd be surprised, probably should live in `fp`.

**Pull request merged**